File: core/src/main/java/io/github/fourlastor/game/coordinates/Hex.java
package io.github.fourlastor.game.coordinates;

import com.badlogic.gdx.math.GridPoint2;
import com.badlogic.gdxplus.math.GridPoint3;
import java.util.Arrays;
import java.util.Optional;

/**
 * Hex coordinates, in offset (odd-q) and cube systems.
 * See <a href="https://www.redblobgames.com/grids/hexagons/#coordinates">original blog post</a>.
 * Cube is useful for pathfinding, offset is what we get from tiled, and it's useful for the on-screen coordinates.
 */
public class Hex {

    public final GridPoint2 offset;
    public final GridPoint3 cube;

    public Hex(GridPoint2 offset) {
        this.offset = offset;
        this.cube = toCube(offset);
    }

    public Hex(GridPoint3 cube) {
        this.cube = cube;
        this.offset = toOffset(cube);
    }

    public static GridPoint3 toCube(GridPoint2 offset) {
        int q = offset.x;
        int r = offset.y - (offset.x - (Math.abs(offset.x % 2))) / 2;
        return new GridPoint3(q, r, -q - r);
    }

    public static GridPoint2 toOffset(GridPoint3 cube) {

        int x = cube.x;
        int y = cube.y + (cube.x - (Math.abs(cube.x % 2))) / 2;
        return new GridPoint2(x, y);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Hex hex = (Hex) o;

        if (!offset.equals(hex.offset)) return false;
        return cube.equals(hex.cube);
    }

    @Override
    public int hashCode() {
        int result = offset.hashCode();
        result = 31 * result + cube.hashCode();
        return result;
    }

    public void set(Hex other) {
        offset.set(other.offset);
        cube.set(other.cube);
    }

    public boolean sameAxisAs(Hex hex) {
        return hex.cube.x == cube.x || hex.cube.y == cube.y || hex.cube.z == cube.z;
    }

    public Hex offset(Direction direction, int amount) {
        return new Hex(cube.cpy().add(direction.cube.cpy().scl(amount)));
    }

    public enum Direction {
        S(0, -1, 1),
        SE(1, -1, 0),
        NE(1, 0, -1),
        N(0, +1, -1),
        NW(-1, 1, 0),
        SW(-1, 0, 1);

        private final GridPoint3 cube;

        Direction(int x, int y, int z) {
            cube = new GridPoint3(x, y, z);
        }

        public static Direction byValue(GridPoint3 cube) {
            return Arrays.stream(values())
                    .filter(dir -> dir.cube.equals(cube))
                    .findFirst()
                    .get();
        }

        public static Direction byName(String value) {

            Optional<Direction> test = Arrays.stream(values())
                    .filter(dir -> dir.name().equalsIgnoreCase(value))
                    .findFirst();
            return test.orElse(null);
        }

        public Direction opposite() {
            return byValue(cube.cpy().inverse());
        }

        /**
         * Starting at 0 degrees = East direction, counter-clockwise.
         */
        public static Direction fromRotation(int degrees) {

            degrees = Math.floorMod(degrees, 360);
            if (degrees <= 60) return NE;
            else if (degrees <= 120) return N;
            else if (degrees <= 180) return NW;
            else if (degrees <= 240) return SW;
            else if (degrees <= 300) return S;
            else if (degrees <= 360) return SE;
            return null;
        }
    }
}

File: core/src/main/java/io/github/fourlastor/game/coordinates/Hex.kt
package io.github.fourlastor.game.coordinates

import com.badlogic.gdx.math.GridPoint2
import com.badlogic.gdx.math.GridPoint3
import io.github.fourlastor.game.extensions.GridPoint3s.inverse
import io.github.fourlastor.game.extensions.GridPoint3s.scl
import kotlin.math.abs

/**
 * Hex coordinates, in offset (odd-q) and cube systems.
 * See [original blog post](https://www.redblobgames.com/grids/hexagons/#coordinates).
 * Cube is useful for pathfinding, offset is what we get from tiled, and it's useful for the on-screen coordinates.
 */

data class Hex @JvmOverloads constructor(
    val offset: GridPoint2,
    val cube: GridPoint3 = toCube(offset),
) {

    fun set(other: Hex) {
        offset.set(other.offset)
        cube.set(other.cube)
    }

    fun sameAxisAs(hex: Hex): Boolean {
        return hex.cube.x == cube.x || hex.cube.y == cube.y || hex.cube.z == cube.z
    }

    fun offset(direction: Direction, amount: Int): Hex {
        return Hex(toOffset(cube.cpy().add(direction.cube.cpy().scl(amount))))
    }

    enum class Direction(x: Int, y: Int, z: Int) {
        S(0, -1, 1),
        SE(1, -1, 0),
        NE(1, 0, -1),
        N(0, +1, -1),
        NW(-1, 1, 0),
        SW(-1, 0, 1);

        val cube: GridPoint3 = GridPoint3(x, y, z)

        fun opposite(): Direction = byValue(cube.cpy().inverse())

        companion object {
            fun byValue(cube: GridPoint3): Direction = values().first { it.cube == cube }

            fun byName(value: String): Direction = values().first { it.name.equals(value, ignoreCase = true) }

            /**
             * Starting at 0 degrees = East direction, counter-clockwise.
             */
            fun fromRotation(degrees: Int): Direction = Math.floorMod(degrees, 360).let {
                when {
                    it <= 60 -> NE
                    it <= 120 -> N
                    it <= 180 -> NW
                    it <= 240 -> SW
                    it <= 300 -> S
                    else -> SE
                }
            }
        }
    }

    companion object {
        fun toCube(offset: GridPoint2): GridPoint3 {
            val q = offset.x
            val r = (offset.y - (offset.x - abs((offset.x % 2).toDouble())) / 2).toInt()
            return GridPoint3(q, r, -q - r)
        }

        fun toOffset(cube: GridPoint3): GridPoint2 {
            val x = cube.x
            val y = (cube.y + (cube.x - abs((cube.x % 2).toDouble())) / 2).toInt()
            return GridPoint2(x, y)
        }
    }
}
